### DISCUSSION TITLE
ParameterBlock set shifted when using Push Constants?

### DISCUSSION THREAD

**Discussion by fnuecke** [shader-slang/slang#5863] [Q&A] [answered]
I'm still very new to slang, so I may be doing something wrong, but I'm observing an unexpected shift in binding sets when using push constants. 

Here's a small shader [demonstrating this in the playground](https://shader-slang.com/slang-playground/?target=SPIRV&code=eJwtTztrwzAQ3vUrjkw2BCWBbk4zBK-BQkeT4ZBk-4j14CSFpKX_vbKSu5u-F9_FxFkl6DEh_Aoo812RzEaf8zgaPmZy6eMEukg68deJL2S0Jhk-L17djqv1BFNfaTGxzyHOWOywGiHgc_GoCzUUWBtuNmjDQiMpTOTdpr2KwWWbZjaoY3PYli3Y3ZOGhPF2QXJNdjR6tq9IbMWr6m4HWHU-JLL0UwOBjfV3chOgUiZGKWUVv4vAJ-y7CvQUAyY1X0ycm9pfrj8O-6t8bOFQ721qy-PiH5v9YqQ).

Note the `OpDecorate %gData_data DescriptorSet 1`.  I'd expect this to be `0`.  Removing the `uniform uint a` from the function signature, i.e. the push constants, results in the expected `OpDecorate %gData_data DescriptorSet 0`.

Am I doing something wrong, misunderstanding something, or could this be a bug in how sets are allocated/assigned?

I did some searching, the most relevant thing I could find was [this issue](https://github.com/shader-slang/slang/issues/5595), but it's about specialization constants instead of push constants, and I'm still seeing this on 2024.17.

For now I'm working around this by not using ParameterBlock and manually defining the layout, which works fine.

**Comment by kaizhangNV** [comment]
I think this is probably a bug in the descriptor set allocation logic, but I'm not sure.
@csyonghe to answer this.

**Reply by csyonghe** [reply, marked as answer]
Yes this is a bug, we should convert it into an issue and fix it, similar to how we fixed the specialization constant case.

**Comment by kaizhangNV** [comment]
Filed an issue for this https://github.com/shader-slang/slang/issues/5865.

**Comment by fnuecke** [comment]
Thanks for the quick response and confirming!
